### PR TITLE
chore(gatsby): convert build-html to typescript

### PR DESCRIPTION
**Pull request by blainekasten**
Converts `commands/build-html` to typescript. Also moves some shared types in the commands folder into a relative `/types` folder. There is likely a use-case to move the shared types even further up, but I'm thinking we'll wait until that structure becomes more obvious

### CODE REVIEW

File: jest-transformer.js [hunk 1]
```diff
@@ -1,2 +1,11 @@
 const babelPreset = require(`babel-preset-gatsby-package`)()
-module.exports = require(`babel-jest`).createTransformer(babelPreset)
+module.exports = require(`babel-jest`).createTransformer({
```
Comment (acao) [inline jest-transformer.js]: is there a reason to not use the `ts-jest` transformer directly? I can imagine a number of reasons why, just curious

Comment (blainekasten) [reply jest-transformer.js]: I can't really explain it. But ts-jest wasn't allowing us to mix commonjs and module import/exports. Builds would be fine and all, but tests would break and basically complain that things didn't exist because the require didn't resolve to the same as the export.

And honestly this makes more sense anyhow. We are using babel for js files, we are using babel for ts files, why not use it for tests too.

File: jest.config.js [hunk 2]
```diff
@@ -35,7 +35,7 @@ module.exports = {
   ],
   transform: {
     "^.+\\.js$": `<rootDir>/jest-transformer.js`,
-    "^.+\\.tsx?$": `<rootDir>/node_modules/ts-jest/preprocessor.js`,
+    "^.+\\.tsx?$": `<rootDir>/jest-transformer.js`,
```
Comment (ascorbic) [inline jest.config.js]: As per my comment, these could be combined

File: packages/gatsby/src/commands/build-html.ts [hunk 3]
```diff
@@ -1,18 +1,21 @@
...
-// const convertHrtime = require(`convert-hrtime`)
-const { chunk } = require(`lodash`)
-const webpackConfig = require(`../utils/webpack.config`)
-const reporter = require(`gatsby-cli/lib/reporter`)
-const { createErrorFromString } = require(`gatsby-cli/lib/reporter/errors`)
-const telemetry = require(`gatsby-telemetry`)
+import Bluebird from "bluebird"
+import webpack from "webpack"
+import fs from "fs-extra"
+// import convertHrtime from "convert-hrtime"
```
Comment (ascorbic) [inline packages/gatsby/src/commands/build-html.ts]: Is there a reason you've left this in?

File: packages/gatsby/src/commands/build-html.ts [hunk 4]
```diff
@@ -1,18 +1,21 @@
...
-const Promise = require(`bluebird`)
-const webpack = require(`webpack`)
-const fs = require(`fs-extra`)
-// const convertHrtime = require(`convert-hrtime`)
-const { chunk } = require(`lodash`)
-const webpackConfig = require(`../utils/webpack.config`)
-const reporter = require(`gatsby-cli/lib/reporter`)
-const { createErrorFromString } = require(`gatsby-cli/lib/reporter/errors`)
-const telemetry = require(`gatsby-telemetry`)
+import Bluebird from "bluebird"
```
Comment (ascorbic) [inline packages/gatsby/src/commands/build-html.ts]: Can't this just be imported as `Promise`?

File: packages/gatsby/src/commands/build-html.ts [hunk 5]
```diff
@@ -22,7 +25,10 @@ const runWebpack = compilerConfig =>
     })
   })
 
-const doBuildRenderer = async (program, webpackConfig) => {
+const doBuildRenderer = async (
+  program: IProgram,
+  webpackConfig: webpack.Configuration
+): Promise<string> => {
```
Comment (ascorbic) [inline packages/gatsby/src/commands/build-html.ts]: As a rule, are we going to always declare the return type even when it can be inferred?

File: packages/gatsby/src/commands/build-html.ts [hunk 6]
```diff
@@ -35,15 +41,19 @@ const doBuildRenderer = async (program, webpackConfig) => {
   return outputFile
 }
 
-const buildRenderer = async (program, stage, { parentSpan }) => {
+const buildRenderer = async (
+  program: IProgram,
+  stage: BuildHTMLStage,
+  parentSpan: IActivity
```
Comment (ascorbic) [inline packages/gatsby/src/commands/build-html.ts]: `parentSpan` is destructured in the previous code. Was that a mistake? It does seem a bit odd to destructure it then immediately recreate the object below.

File: packages/gatsby/src/commands/build-html.ts [hunk 7]
```diff
@@ -35,15 +41,19 @@ const doBuildRenderer = async (program, webpackConfig) => {
...
+): Promise<string> => {
   const { directory } = program
   const config = await webpackConfig(program, directory, stage, null, {
     parentSpan,
```
Comment (ascorbic) [inline packages/gatsby/src/commands/build-html.ts]: As per the previous comment, this has changed from the previous code.

File: packages/gatsby/src/commands/build-html.ts [hunk 8]
```diff
@@ -53,27 +63,28 @@ const deleteRenderer = async rendererPath => {
...
 
-    Promise.map(
+    Bluebird.map(
       segments,
       pageSegment =>
-        new Promise((resolve, reject) => {
+        new Bluebird<void>((resolve, reject) => {
           workerPool
             .renderHTML({
               htmlComponentRendererPath,
```
Comment (ascorbic) [inline packages/gatsby/src/commands/build-html.ts]: Could we use async/await here? (I realise this is old code)

File: packages/gatsby/src/commands/build-html.ts [hunk 9]
```diff
@@ -95,22 +106,22 @@ const renderHTMLQueue = (
             .catch(reject)
         })
     )
-      .then(resolve)
+      .then(() => resolve())
```
Comment (ascorbic) [inline packages/gatsby/src/commands/build-html.ts]: Why did you need to change this?

File: packages/gatsby/src/commands/build-html.ts [hunk 10]
```diff
@@ -127,14 +138,18 @@ const buildPages = async ({
...
+  activity: IActivity
+  workerPool: IWorkerPool
+}): Promise<void> => {
+  const rendererPath = await buildRenderer(program, stage, activity.span)
+  await doBuildPages(rendererPath, pagePaths, activity, workerPool)
   await deleteRenderer(rendererPath)
 }
 
-module.exports = {
+export default {
```
Comment (ascorbic) [inline packages/gatsby/src/commands/build-html.ts]: This feels ~~a bit icky~~ non-idiomatic. If you change this to a named export, you could just do a named import in build.ts and develop.ts.

File: packages/gatsby/src/commands/build.js [hunk 11]
```diff
@@ -2,7 +2,7 @@
 
 const path = require(`path`)
 const report = require(`gatsby-cli/lib/reporter`)
-const buildHTML = require(`./build-html`)
+import buildHTML from "./build-html"
```
Comment (ascorbic) [inline packages/gatsby/src/commands/build.js]: As per above comment, this could be changed to named import and export.

File: packages/gatsby/src/commands/build-html.ts (hunk 3 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/build-html.ts]: It was in there before, so I figured it was left for a reason /shrug

File: packages/gatsby/src/commands/build-html.ts (hunk 5 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/build-html.ts]: Yeah I think so. Easier to visually grok and we have it set as a lint rule

File: packages/gatsby/src/commands/build-html.ts (hunk 9 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/build-html.ts]: Because typescript caught a potential bug. This code is intended to resolve<void>, but it was actually resolving with the value of the above promise's. Which isn't void.

File: packages/gatsby/src/commands/build.js (hunk 11 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/build.js]: Yeah. I agree with this. I'll do that :)

File: packages/gatsby/src/commands/build-html.ts (hunk 4 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/build-html.ts]: Typescript was throwing an error that we were overriding a global Promise. So I think we'll have to keep doing this

File: packages/gatsby/src/commands/build-html.ts (hunk 8 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/build-html.ts]: Yeah! Good call out. I improved a few paths that looked like async/await was better

File: packages/gatsby/src/commands/build-html.ts (hunk 7 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/build-html.ts]: I don't actually know the context around this, but I see it everywhere. Just wrapping in an object doesn't change any references, so we are still always using the same instance all around. I'm going to start removing these blocks as I can.

Comment (blainekasten) [reply packages/gatsby/src/commands/build-html.ts]: asking in this thread. https://gatsbyjs.slack.com/archives/CA1GW1HNU/p1579121474098900

File: packages/gatsby/src/commands/build-html.ts (hunk 6 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/build-html.ts]: https://gatsbyjs.slack.com/archives/CA1GW1HNU/p1579121474098900

File: packages/gatsby/src/commands/build-html.ts [hunk 12]
```diff
@@ -132,7 +120,7 @@ const doBuildPages = async (
   }
 }
 
-const buildPages = async ({
+const buildHTML = async ({
```
Comment (ascorbic) [inline packages/gatsby/src/commands/build-html.ts]: Rather than a default export, how about a named export here? i.e. `export const buildPages = ` I generally prefer named over default exports because it helps catch typos. Do you agree?

Comment (blainekasten) [reply packages/gatsby/src/commands/build-html.ts]: I've been back and forth on this. I do like named exports for that reason. The reason i've used export default is mostly so future devs think twice about exporting other methods. Not speaking about Gatsby devs per se, but devs in general tend to not think about API surface and are happy to export anything that "gets a job done."

I realize export default doesn't really solve this. But might lend in a small way to prevent api surface creep

File: jest-transformer.js (hunk 1 above)
Comment (acao) [reply jest-transformer.js]: nice, that makes sense. eventually using project references and the module name mapper, you could build all typescript projects at once using `tsc —build` which is the official typescript reccomendation for large, multi-module projects. ts-jest can honor this if you tweak moduleNameMapper right, which means you can execute jest from the root and resolve all local interdependencies without running `tsc` at all.

Comment (blainekasten) [reply jest-transformer.js]: Yes! That is potentially a  goal once we achieve 100% transition

File: packages/gatsby/src/commands/build-html.ts (hunk 4 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/build-html.ts]: Yup. I think i swapped all the regular promises out below for async/await

File: packages/gatsby/src/commands/build-html.ts (hunk 12 above)
Comment (ascorbic) [reply packages/gatsby/src/commands/build-html.ts]: I agree that we want to avoid API creep, but there's no specific reason to only export one per file. It would make sense to agree a style guide for this sort of thing so we can be consistent

File: packages/gatsby/src/redux/types.ts [hunk 13]
```diff
@@ -1,7 +1,23 @@
+export enum ProgramStatus {
+  BOOTSTRAP_FINISHED = `BOOTSTRAP_FINISHED`,
+  BOOTSTRAP_QUERY_RUNNING_FINISHED = `BOOTSTRAP_QUERY_RUNNING_FINISHED`,
+}
+
 export interface IReduxState {
-  nodes?: any // TODO
-  nodesByType: Map<any, any> // TODO
+  status: ProgramStatus
+  nodes?: {
```
Comment (ascorbic) [inline packages/gatsby/src/redux/types.ts]: If you extract this into an interface, you can then easily type nodesByType below

File: packages/gatsby/src/commands/build-html.ts (hunk 12 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/build-html.ts]: I read into this more and i'm not pretty convinced the default export was a poor design for es modules.. happy to switch this to named exports

File: packages/gatsby/src/redux/types.ts (hunk 13 above)
Comment (blainekasten) [reply packages/gatsby/src/redux/types.ts]: sounds great, i'll follow that up in a future PR. I didn't fully understand yet what `nodesByType` was
